python/tank/platform/interactive_authentication.py

```diff
@@ -14,172 +14,272 @@
 from tank.util import authentication
 from tank.util.login import get_login_name
 from tank.util import shotgun
 
 # Configure logging
 logger = logging.getLogger("sgtk.interactive_authentication")
-# logger.setLevel(logging.DEBUG)
+logger.setLevel(logging.DEBUG)
 logger.addHandler(logging.StreamHandler())
 
 
-def _get_session_token(hostname, login, password, http_proxy):
+class AuthenticationHandlerBase(object):
     """
-    Retrieves a session token for the given credentials.
-    :param hostname: The host to connect to.
-    :param login: The user to get a session for.
-    :param password: Password for the user.
-    :param http_proxy: Proxy to use. Can be None.
-    :returns: If the credentials were valid, returns a session token, otherwise returns None.
+    Base class for authentication requests. It handles locking reading cached credentials
+    on disk and writing newer credentials back. It also keeps track of any attempt to cancel
+    authentication.
     """
-    try:
-        return authentication.generate_session_token(hostname, login, password, http_proxy)
-    except TankAuthenticationError:
-        print "Authentication failed."
-        return None
 
-
-def _do_ui_based_login(hostname, login, http_proxy):
-    pass
-
-
-def _get_keyboard_input(label, default_value=""):
+    _authentication_lock = threading.Lock()
     """
-    Queries for keyboard input.
-    :param label: The name of the input we require.
-    :param default_value: The value to use if the user has entered no input.
-    :returns: The user input or default_value if nothing was entered.
+    Lock the assures only one thread at a time can execute the authentication logic.
     """
-    text = label
-    if default_value:
-        text += ' [Default: %s]' % default_value
-    text += ": "
-    user_input = None
-    while not user_input:
-        user_input = raw_input(text) or default_value
-    return user_input
-
-
-def _get_user_credentials_from_keyboard(hostname, login):
+    _authentication_disabled = False
     """
-    Reads the user credentials from the keyboard.
-    :param hostname: Name of the host we will be logging on.
-    :param login: Default value for the login.
-    :returns: A tuple of (login, password) strings.
+    Flag that keeps track if a user cancelled authentication. When the flag is raised, it will
+    be impossible to authenticate again.
     """
-    print "Please enter your login credentials for %s" % hostname
-    login = _get_keyboard_input("Login", login)
-    password = getpass()
-    return login, password
-
 
-def _get_user_pass_from_keyboard(hostname, login):
+    def authenticate(self):
+        """
+        Common login logic, regardless of how we are actually logging in. It will first try to reuse
+        any existing session and if that fails then it will ask for credentials and upon success
+        the credentials will be cached.
+        """
+        with AuthenticationHandlerBase._authentication_lock:
```

Comment:
cool! Please add a comment too in that case to point this out - I don't we don't use `with` commonly in the core at the moment.